vowpalwabbit/global_data.h

```diff
@@ -33,97 +36,78 @@
 
 struct dictionary_info
 {
-  char* name;
+  std::string name;
   uint64_t file_hash;
-  feature_dict* dict;
+  std::shared_ptr<feature_dict> dict;
 };
 
-inline void deleter(substring ss, uint64_t /* label */) { free_it(ss.begin); }
-
 class namedlabels
 {
  private:
-  std::vector<substring> id2name;
-  v_hashmap<substring, uint64_t> name2id;
-  uint32_t K;
+  // NOTE: This ordering is critical. m_id2name and m_name2id contain pointers into m_label_list!
+  std::string m_label_list;
+  std::vector<VW::string_view> m_id2name;
+  std::unordered_map<VW::string_view, uint64_t> m_name2id;
+  uint32_t m_K;
 
  public:
-  namedlabels(std::string label_list)
+  namedlabels(const std::string& label_list) : m_label_list(label_list)
   {
-    char* temp = calloc_or_throw<char>(1 + label_list.length());
-    memcpy(temp, label_list.c_str(), strlen(label_list.c_str()));
-    substring ss = {temp, nullptr};
-    ss.end = ss.begin + label_list.length();
-    tokenize(',', ss, id2name);
-
-    K = (uint32_t)id2name.size();
-    name2id.delete_v();  // delete automatically allocated vector.
-    name2id.init(4 * K + 1, 0, substring_equal);
-    for (size_t k = 0; k < K; k++)
+    tokenize(',', m_label_list, m_id2name);
+
+    m_K = (uint32_t)m_id2name.size();
+    m_name2id.max_load_factor(0.25);
+    m_name2id.reserve(m_K);
+
+    for (size_t k = 0; k < m_K; k++)
     {
-      substring& l = id2name[k];
-      uint64_t hash = uniform_hash((unsigned char*)l.begin, l.end - l.begin, 378401);
-      uint64_t id = name2id.get(l, hash);
-      if (id != 0)  // TODO: memory leak: char* temp
+      const VW::string_view& l = m_id2name[k];
+      auto iter = m_name2id.find(l);
+      if (iter != m_name2id.end())
         THROW("error: label dictionary initialized with multiple occurances of: " << l);
-      size_t len = l.end - l.begin;
-      substring l_copy = {calloc_or_throw<char>(len), nullptr};
-      memcpy(l_copy.begin, l.begin, len * sizeof(char));
-      l_copy.end = l_copy.begin + len;
-      name2id.put(l_copy, hash, k + 1);
+      m_name2id.emplace(l, k + 1);
     }
   }
 
-  ~namedlabels()
-  {
-    if (id2name.size() > 0)
-      free(id2name[0].begin);
-    name2id.iter(deleter);
-    name2id.delete_v();
-  }
+  uint32_t getK() { return m_K; }
 
-  uint32_t getK() { return K; }
-
-  uint64_t get(substring& s)
+  uint64_t get(VW::string_view s) const
   {
-    uint64_t hash = uniform_hash((unsigned char*)s.begin, s.end - s.begin, 378401);
-    uint64_t v = name2id.get(s, hash);
-    if (v == 0)
+    auto iter = m_name2id.find(s);
+    if (iter == m_name2id.end())
     {
-      std::cerr << "warning: missing named label '";
-      for (char* c = s.begin; c != s.end; c++) std::cerr << *c;
-      std::cerr << '\'' << std::endl;
+      std::cerr << "warning: missing named label '" << s << '\'' << std::endl;
+      return 0;
     }
-    return v;
+    return iter->second;
   }
 
-  substring get(uint32_t v)
+  VW::string_view get(uint32_t v) const
```

Comment:
Can you make the two get functions use consistently either u32 or u64?

Reply:
Technically these should be size_t's since they're indices in a vector. But callers of these functions are using uint32_t's. Could possibly lead to issues someday, but I doubt we'll ever have more than ~4 Billion label names or run on an environment with a 16-bit size_t

Reply:
That second one I could see happening. Could you add a static assert for sizeof(size_t) >= sizeof(uint32_t), or similar to call out the issue?